[FLAX] Question Answering Example

kamalkraj opened the pull request:
# What does this PR do?

Flax Question Answering Example 

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->
@patrickvonplaten @patil-suraj @sgugger

kamalkraj committed:
flax qa example

sgugger commented on examples/flax/question-answering/README.md:
````diff
@@ -0,0 +1,100 @@
+You can do this either directly on [huggingface.co](https://huggingface.co/new) (assuming that
+you are logged in) or via the command line:
+
+```
+huggingface-cli repo create bert-qa-squad-test
+```
+
+Next we clone the model repository to add the tokenizer and model files.
+
+```
+git clone https://huggingface.co/<your-username>/bert-qa-squad-test
+```
````
Not linked to this PR per se, but this is not necessary while using the `Repository` API and all Flax examples should be updated to use it, while removing those instructions.

sgugger commented on examples/flax/question-answering/README.md:
```diff
@@ -0,0 +1,100 @@
+
+
+### Usage notes
+Note that when contexts are long they may be split into multiple training cases, not all of which may contain
```
```suggestion

Note that when contexts are long they may be split into multiple training cases, not all of which may contain
```

sgugger commented on examples/flax/question-answering/utils_qa.py:
```diff
@@ -0,0 +1,427 @@
+# coding=utf-8
```
If this is a copy from the PyTorch example folder, it should be added [here](https://github.com/huggingface/transformers/blob/04976a32dc555667afa994e8f918cbee88d84a4f/utils/check_copies.py#L31) so the copy is checked by our internal tooling.

kamalkraj replied:
Done

sgugger reviewed (approved):
Thanks for adding that example!

kamalkraj committed:
Updated README:  Added Large model

kamalkraj committed:
added utils_qa.py FULL_COPIES

patil-suraj reviewed (approved):
Looks really nice! Thanks for adding this example.

patil-suraj commented on examples/flax/question-answering/run_qa.py:
```diff
@@ -0,0 +1,893 @@
+        step = 0
+        # create a numpy array and fill it with -100.
+        logits_concat = np.full((len(dataset), max_len), -100, dtype=np.float64)
+        # Now since we have create an array now we will populate it with the outputs gathered using accelerator.gather
```
this comment should be adapted slightly, since we don't use `accelerator` here

kamalkraj committed:
Updates:
1. Copyright Year updated
2. added dtype arg
3. passing seed and dtype to load model
4. Check eval flag before running eval

kamalkraj committed:
updated README

kamalkraj commented on examples/flax/question-answering/run_qa.py:
```diff
@@ -0,0 +1,893 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2020 The HuggingFace Team All rights reserved.
```
done

kamalkraj commented on examples/flax/question-answering/run_qa.py:
```diff
@@ -0,0 +1,893 @@
+        model_args.model_name_or_path,
+        config=config,
+        cache_dir=model_args.cache_dir,
+        revision=model_args.model_revision,
+        use_auth_token=True if model_args.use_auth_token else None,
```
done.

kamalkraj replied:
@patil-suraj 
Why there is no Dynamic loss scaling for dtype `float16` ? 

 https://flax.readthedocs.io/en/latest/_autosummary/flax.optim.DynamicScale.html#flax.optim.DynamicScale

cc @patrickvonplaten

patil-suraj replied:
After the #13098, `dtype` will specify only the `dtype` of computation, so just the forward pass will be in half-precision and grads will still be computed in `fp32` so for this case loss scaling is not required.

True mixed-precision training for flax examples is planned and will include loss scaling, will post more about it soon :) 

Also, note that we don't use flax optimizers here, we are using `optax`.

kamalkraj commented on examples/flax/question-answering/run_qa.py:
```diff
@@ -0,0 +1,893 @@
+                train_metrics = []
+
+            if (cur_step % training_args.eval_steps == 0 or cur_step % step_per_epoch == 0) and cur_step > 0:
+
```
done

kamalkraj committed:
updated code comment

kamalkraj commented:
@patil-suraj 
Done changes according to your review.
